core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java

```diff
@@ -0,0 +1,108 @@
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap;
+import java.util.Set;
+import java.util.UUID;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The first node list refresh: contact points are not in the metadata yet, we need to copy them
+ * over.
+ */
+@ThreadSafe
+class InitialNodeListRefresh extends NodesRefresh {
+
+  private static final Logger LOG = LoggerFactory.getLogger(InitialNodeListRefresh.class);
+
+  @VisibleForTesting final Iterable<NodeInfo> nodeInfos;
+  @VisibleForTesting final Set<Node> contactPoints;
+
+  InitialNodeListRefresh(Iterable<NodeInfo> nodeInfos, Set<Node> contactPoints) {
+    this.nodeInfos = nodeInfos;
+    this.contactPoints = contactPoints;
+  }
+
+  @Override
+  public Result compute(
+      DefaultMetadata oldMetadata, boolean tokenMapEnabled, InternalDriverContext context) {
+
+    String logPrefix = context.getSessionName();
+    TokenFactoryRegistry tokenFactoryRegistry = context.getTokenFactoryRegistry();
+
+    assert oldMetadata.getNodes().isEmpty();
+
+    TokenFactory tokenFactory =
+        oldMetadata.getTokenMap().map(m -> ((DefaultTokenMap) m).getTokenFactory()).orElse(null);
+    boolean tokensChanged = false;
+
+    ImmutableMap.Builder<UUID, Node> newNodesBuilder = ImmutableMap.builder();
+
+    for (NodeInfo nodeInfo : nodeInfos) {
+      EndPoint endPoint = nodeInfo.getEndPoint();
+      DefaultNode node = findIn(contactPoints, endPoint);
+      if (node == null) {
+        node = new DefaultNode(endPoint, context);
+        LOG.debug("[{}] Adding new node {}", logPrefix, node);
+      } else {
+        LOG.debug("[{}] Copying contact point {}", logPrefix, node);
+      }
+      if (tokenFactory == null && nodeInfo.getPartitioner() != null) {
+        tokenFactory = tokenFactoryRegistry.tokenFactoryFor(nodeInfo.getPartitioner());
+      }
+      tokensChanged |= copyInfos(nodeInfo, node, tokenFactory, logPrefix);
+      newNodesBuilder.put(node.getHostId(), node);
+    }
+
+    ImmutableMap<UUID, Node> newNodes = newNodesBuilder.build();
+    ImmutableList.Builder<Object> eventsBuilder = ImmutableList.builder();
+
+    for (Node newNode : newNodes.values()) {
+      if (!contactPoints.contains(newNode)) {
+        eventsBuilder.add(NodeStateEvent.added((DefaultNode) newNode));
+      }
+    }
+    for (Node contactPoint : contactPoints) {
+      if (findIn(newNodes.values(), contactPoint.getEndPoint()) == null) {
+        eventsBuilder.add(NodeStateEvent.removed((DefaultNode) contactPoint));
+      }
+    }
+
+    return new Result(
+        oldMetadata.withNodes(newNodes, tokenMapEnabled, tokensChanged, tokenFactory, context),
+        eventsBuilder.build());
+  }
+
+  private DefaultNode findIn(Iterable<Node> nodes, EndPoint endPoint) {
+    for (Node node : nodes) {
+      if (node.getEndPoint().equals(endPoint)) {
+        return (DefaultNode) node;
```

Comment:
Is this cast safe?

Reply:
Yes, the driver doesn't allow you to plug your own `Node` implementation. In this particular case (and a few others), the interface/implementation separation is only meant to hide irrelevant details to the end user.

Reply:
Since `findIn` is private we can also change the signature to `Iterable<DefaultNode>` to avoid the cast, right?